lib/core/src/Cardano/Pool/DB.hs

```diff
@@ -166,6 +190,98 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
         -- For a Sqlite DB, this would be "run a query inside a transaction".
     }
 
+-- | Represents an abstract notion of a certificate publication time.
+--
+-- Certificates published at later times take precedence over certificates
+-- published at earlier times.
+--
+type CertificatePublicationTime = (SlotId, SlotInternalIndex)
+
+-- | Indicates the current registration status of a pool.
+--
+-- Use the 'readPoolRegistrationStatus' function to query the registration
+-- status for a particular pool and database backend.
+--
+data PoolRegistrationStatus
+    = PoolNotRegistered
+        -- ^ Indicates that a pool is not registered.
+    | PoolRegistered
+        PoolRegistrationCertificate
+        -- ^ Indicates that a pool is registered BUT NOT marked for retirement.
+        -- Records the latest registration certificate.
+    | PoolRegisteredAndRetired
+        PoolRegistrationCertificate
+        PoolRetirementCertificate
+        -- ^ Indicates that a pool is registered AND ALSO marked for retirement.
+        -- Records the latest registration and retirement certificates.
+    deriving (Eq, Show)
+
+-- | Given the latest registration and retirement certificates for a pool,
+-- determine the pool's current registration status based on the relative
+-- order in which the certificates were published.
+determinePoolRegistrationStatus
+    :: Ord certificatePublicationTime
+    => Maybe (certificatePublicationTime, PoolRegistrationCertificate)
+    -> Maybe (certificatePublicationTime, PoolRetirementCertificate)
+    -> PoolRegistrationStatus
+determinePoolRegistrationStatus = f
```

Comment:
> I'm pretty sure this could be made shorter with `case` and guards, without the need for helper functions f and g.

It does indeed save a few lines (8 lines added in return for 12 lines deleted).

I've fixed this in c11624eb76afacf8448f279656da74b940f8b443.